File: android/src/main/java/com/lightningkite/khrysalis/observables/StandardObservableProperty.shared.kt
package com.lightningkite.khrysalis.observables

import com.lightningkite.khrysalis.Box
import com.lightningkite.khrysalis.boxWrap
import io.reactivex.subjects.PublishSubject
import io.reactivex.subjects.Subject

class StandardObservableProperty<T>(
    var underlyingValue: T,
    override val onChange: Subject<Box<T>> = PublishSubject.create()
) : MutableObservableProperty<T>() {
    override var value: T
        get() = underlyingValue
        set(value) {
            underlyingValue = value
            onChange.onNext(boxWrap(value))
        }

    override fun update() {
        onChange.onNext(boxWrap(value))
    }
}

File: ios/Khrysalis/com/lightningkite/khrysalis/observables/StandardObservableProperty.shared.swift
//Package: com.lightningkite.khrysalis.observables
//Converted using Khrysalis2

import Foundation
import RxSwift
import RxRelay



public class StandardObservableProperty<T>: MutableObservableProperty<T> {
    
    public var underlyingValue: T
    public var _onChange: Subject<T>
    override public var onChange: Subject<T> { get { return _onChange } set(value) { _onChange = value } }
    
    override public var value: T {
        get {
            return underlyingValue
        }
        set(value) {
            underlyingValue = value
            onChange.onNext(value)
        }
    }
    
    override public func update() -> Void {
        onChange.onNext(value)
    }
    
    public init(underlyingValue: T, onChange: Subject<T> = PublishSubject.create()) {
        self.underlyingValue = underlyingValue
        self._onChange = onChange
        super.init()
    }
    convenience public init(_ underlyingValue: T, _ onChange: Subject<T> = PublishSubject.create()) {
        self.init(underlyingValue: underlyingValue, onChange: onChange)
    }
}
 
